[FLOC-4013] Add a script that can help with merging a pull request.

james-w opened the pull request:
admin/merge_pr https://github.com/ClusterHQ/flocker/pull/<id>

will loop until all the checks are green, then trigger a merge
and delete the branch.

Rather than having to decide between merging with failing tests,
or spending time triggering retries and checking status, you can
use this script to do the work. Give it the URL of the pull request
and it will loop checking the status. When it is green it will
merge the PR, delete the source branch from github and exit.

If tests fail then it will retry the Jenkins job once, and if they
fail again it will tell you about it so that you can take a look.
If the test failures aren't your fault you can click retry and it
will merge if it is green from that run. If they are your fault
then kill the script and make the fixes.

If you just want to get a green run before asking for a review
then use --no-merge, it will exit when the tests are green without
triggering the merge.

You could combine the script with "undistract-me" or "notify-send"
to get a popup when the tests are green.

wallrj commented on admin/merge_pr:
```diff
@@ -0,0 +1,401 @@
+            "You must set {} and {}.\n"
+            "You can generate a token at "
+            "https://github.com/settings/tokens\n".format(
+                GITHUB_USERNAME_ENV_VAR, GITHUB_TOKEN_ENV_VAR))
```
Add a note about which "scopes" to assign to the token.

wallrj commented:
Thanks @james-w 

Looks good so far.

* The output is a little verbose:

```
Can't merge PR yet because these 8 checks haven't succeeded:
jenkins-run_trial_on_AWS_CentOS_7_flocker.node.functional.test_docker: running (http://ci-live.clusterhq.com:8080/job/ClusterHQ-flocker/job/count-cloudformation-success-FLOC-4003/job/run_trial_on_AWS_CentOS_7_flocker.node.functional.test_docker/)
jenkins-run_acceptance_loopback_on_AWS_CentOS_7_for_flocker.acceptance: running (http://ci-live.clusterhq.com:8080/job/ClusterHQ-flocker/job/count-cloudformation-success-FLOC-4003/job/run_acceptance_loopback_on_AWS_CentOS_7_for_flocker.acceptance/)
jenkins-run_acceptance_loopback_on_AWS_Ubuntu_Trusty_for_flocker.acceptance: running (http://ci-live.clusterhq.com:8080/job/ClusterHQ-flocker/job/count-cloudformation-success-FLOC-4003/job/run_acceptance_loopback_on_AWS_Ubuntu_Trusty_for_flocker.acceptance/)
jenkins-run_trial_for_ebs_storage_driver_on_CentOS_7_flocker.node.agents.functional: running (http://ci-live.clusterhq.com:8080/job/ClusterHQ-flocker/job/count-cloudformation-success-FLOC-4003/job/run_trial_for_ebs_storage_driver_on_CentOS_7_flocker.node.agents.functional/)
jenkins-run_trial_on_AWS_CentOS_7_as_root_flocker.node.agents: success (http://ci-live.clusterhq.com:8080/job/ClusterHQ-flocker/job/count-cloudformation-success-FLOC-4003/job/run_trial_on_AWS_CentOS_7_as_root_flocker.node.agents/)
jenkins-run_trial_for_ebs_storage_driver_on_Ubuntu_trusty_flocker.node.agents.functional: running (http://ci-live.clusterhq.com:8080/job/ClusterHQ-flocker/job/count-cloudformation-success-FLOC-4003/job/run_trial_for_ebs_storage_driver_on_Ubuntu_trusty_flocker.node.agents.functional/)
jenkins-run_trial_for_cinder_storage_driver_on_CentOS_7_flocker.node.agents.functional: running (http://ci-live.clusterhq.com:8080/job/ClusterHQ-flocker/job/count-cloudformation-success-FLOC-4003/job/run_trial_for_cinder_storage_driver_on_CentOS_7_flocker.node.agents.functional/)
jenkins-run_trial_for_cinder_storage_driver_on_Ubuntu_trusty_flocker.node.agents.functional: running (http://ci-live.clusterhq.com:8080/job/ClusterHQ-flocker/job/count-cloudformation-success-FLOC-4003/job/run_trial_for_cinder_storage_driver_on_Ubuntu_trusty_flocker.node.agents.functional/)
Sleeping for 60 seconds and trying again.

```

I'd prefer to see only the URLs

* The ``--help`` doesn't say anything about the environment variables and the error messages when those variables are missing don't give enough information about the github token scope requirements.

* I tried it on a PR where there were no existing Jenkins builds....I think because @sarum90 had just restarted Jenkins. It'd be nice if it had triggered a build for me.

I'll wait and see if it successfully merges this PR:
 * https://github.com/ClusterHQ/flocker/pull/2550

wallrj commented:
Ok. All the tests passed and it merged and deleted that branch.

```
(4013)(merge-script-FLOC-4013 ? =)[~/.../HybridLogic/flocker]$ ./admin/merge_pr https://github.com/ClusterHQ/flocker/pull/2550
Attempting merge as 58 checks passed
Pull Request successfully merged at 40fc8d61d1892d1084f556fe20548417e5ac8f17
Branch deleted
```

One problem which can be fixed as a followup is:
```
(4013)(merge-script-FLOC-4013 ? =)[~/.../HybridLogic/flocker]$ ./admin/merge_pr https://github.com/ClusterHQ/flocker/pull/2550
Merge request not open: closed
Sleeping for 60 seconds and trying again.
```

...I guess it should just quit at this point.

Some minor nits:
 * Docstrings don't follow the house style
```
"""
Docstring
"""
```

 * Some of the docstrings are missing parameter and return docs.
 * This script uses argparse here rather than the standard Twisted options that we use elsewhere.
 * All the other `admin/*` scripts are written as stubs importing main functions from a corresponding module so that the functions can be re-used elsewhere.
 * There aren't any tests...what's your excuse :smile:  ?
 * There aren't any docs. Perhaps we can document this if / when it becomes part of our standard development process. Might be worth adding a line or two to the contributing document.

Anyway, it seems really useful tool. I'll start using it and let you know if I find any other problems.

Please answer or address the points above and then merge.

wallrj commented:
Oh one other thing. I like to add some notes when I merge a branch. E.g.

```
[FLOC-4013] Add a script that can help with merging a pull request.

Authors: @wallrj, @myechuri
Reviewer: @sarum60
Fixes: https://clusterhq.atlassian.net/browse/FLOC-4003
```

It'd be nice if it could do that for me.

james-w commented:
Hi,

Thanks for the review.

I added the info about the environment variables and scopes to the
error message and the --help.

Triggering a build when there hasn't been one: that's a little tricky, as it would
involve teaching the script how to find the jobs on Jenkins, and possibly
about the setup job. I'm going to leave this for now, and I think Mark might
be looking at why the jobs sometimes don't get triggered.

Quitting when the PR isn't open: I guess that makes sense. I wrote it to
just loop almost no matter what, but I don't think that's quite right. I think
that's a case where it's far more likely to be a user error than a request
to wait until the PR is re-opened or something.

Docsctring style: fixed

Documenting parameters: I skipped the ones that would duplicate the
textual description, like

 ```python
    """
    Return the path portion of a URL.

    :param unicode url: the URL.
    :return unicode: the path portion of the input URL.
    """
```

There's probably a better way to write this, and the type
is useful information. I have gone through and added the
ones I found.

argparse vs Twisted: yeah, I reached for what I know. I propose
leaving it like this for now.

module: that would be nice, but this is all written such that
it would be trivial to do that, so I suggest we just leave it until
the first time someone wants to re-use anything.

tests: I don't have a good excuse, but I have many weak ones.

docs: yeah, I'll happily document this when you think it makes sense.
I can do it to introduce it to people, or I can leave it to see if anyone
actually uses it. What do you think? I originally just wrote this to see
if it could be done, and then wanted to share it to see if anyone else
thought it was a good idea, but getting it in to the tree is the easiest way
to do that. Maybe we would be better off with a bot which did the
merges for us for instance.

extra commit message: I think that would be reasonable. It
could certainly take a -m and add that, and if we have some
standard information to add like you suggest then it could
be added without too much trouble. I propose leaving that
for a followup though.

GIven that I've made changes, and I am not sure what is the best
thing to do for some of these points, I'd appreciate a second
look at this.

Thanks,

James

wallrj commented:
Thanks @james-w 

The changes look good and I agree that the other changes can be implemented later. 

Two unrelated Jenkins builds have failed so let me test / use the tool to trigger those builds and merge the branch....

wallrj commented:
It could handle unknown PRs better. Another followup:

```
(merge-script-FLOC-4013 ? =)[~/.../HybridLogic/flocker]$ ./admin/merge_pr https://github.com/ClusterHQ/flocker/pull/5000
Traceback (most recent call last):
  File "./admin/merge_pr", line 485, in <module>
    sys.exit(main(sys.argv[1:]))
  File "./admin/merge_pr", line 474, in main
    pr_url, SLEEP_FOR, session, jenkins_session, opts.max_retries)
  File "./admin/merge_pr", line 418, in loop_until_passed
    if pr['state'] != u'open':
KeyError: 'state'
```

wallrj commented on admin/merge_pr:
```diff
@@ -0,0 +1,485 @@
+MAX_RETRIES = 1
+SLEEP_FOR = 60
+
+GITHUB_VARS_REQUIRED = """"You must set {} and {}.
```
Oh, there's one too many `"` here.

wallrj commented on admin/merge_pr:
```diff
@@ -0,0 +1,485 @@
+(this script requires the repo:status and public_repo scopes)
+""".format(GITHUB_USERNAME_ENV_VAR, GITHUB_TOKEN_ENV_VAR)
+
+JENKINS_VARS_REQUIRED = """Please specify the jenkins user and\
```
And the escape is wrong at the end of this line.

wallrj commented on admin/merge_pr:
```diff
@@ -0,0 +1,485 @@
+when all the tests pass.
+
+It requires the following environment variables to be set:
+""" + GITHUB_VARS_REQUIRED + "\n" + JENKINS_VARS_REQUIRED
```
This doesn't read properly:
```
(merge-script-FLOC-4013 ? =)[~/.../HybridLogic/flocker]$ ./admin/merge_pr --help
usage: merge_pr [-h] [--no-merge] [--max-retries MAX_RETRIES] url

Merge a branch when all the tests pass. Given the URL to a pull request on
github this script will loop until all the tests pass, and then merge the
branch. It will retry failures on Jenkins, and will also respond to any
retries you make yourself, and cope with the branch being updated while it is
running, waiting for the tests to pass for the new HEAD. It can optionally
skip the merge if you just want to know when all the tests pass. It requires
the following environment variables to be set: "You must set GITHUB_USER and
GITHUB_TOKEN. You can generate a token at https://github.com/settings/tokens
(this script requires the repo:status and public_repo scopes) Please specify
the jenkins user andpassword in the JENKINS_USER and JENKINS_PASSWORD env
vars.
```

wallrj commented on admin/merge_pr:
```diff
@@ -0,0 +1,485 @@
+
+DESCRIPTION = """Merge a branch when all the tests pass.
+
+Given the URL to a pull request on github this script will
```
github > GitHub

wallrj commented:
Actually I spotted a few typos in the help which you may as well fix before merging. I'll leave it to you.

wallrj commented:
Another success. I used `merge_pr` to merge #2572

```
./admin/merge_pr https://github.com/ClusterHQ/flocker/pull/2572
...
Can't merge PR yet because these 1 checks haven't succeeded:
jenkins-run_trial_on_AWS_Ubuntu_Trusty_flocker.control: running (http://ci-live.clusterhq.com:8080/job/ClusterHQ-flocker/job/packer-parser-FLOC-4050/job/run_trial_on_AWS_Ubuntu_Trusty_flocker.control/)
Sleeping for 60 seconds and trying again.


Attempting merge as 58 checks passed
Pull Request successfully merged at aaccd2928491b265b4bacdd15a2188adf950a650
Branch deleted

```
